File: src/gaia/ui/email_sidecar/fetch.py
# Copyright(C) 2025-2026 Advanced Micro Devices, Inc. All rights reserved.
# SPDX-License-Identifier: MIT
"""Verified binary fetch (port of fetch.ts) — the security boundary.

Resolves the host platform -> looks up the artifact in binaries.lock.json ->
downloads it -> **verifies its SHA-256 against the lock and raises loudly on any
mismatch** -> writes it atomically into the cache -> chmod +x on POSIX. A
tampered/truncated download is rejected before it can ever be spawned. There is
NO 'use it anyway' path.
"""

from __future__ import annotations

import hashlib
import os
import stat
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from gaia.logger import get_logger
from gaia.ui.email_sidecar import platform as plat
from gaia.ui.email_sidecar.errors import IntegrityError, PlatformError

logger = get_logger(__name__)


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def file_sha256(path: Path) -> Optional[str]:
    try:
        return sha256_hex(Path(path).read_bytes())
    except OSError:
        return None


def verify_sha256(data: bytes, expected: str, source_label: str) -> str:
    """Raise ``IntegrityError`` loudly on mismatch — the no-silent-fallback gate."""
    actual = sha256_hex(data)
    if actual.lower() != expected.lower():
        raise IntegrityError(
            f"SHA-256 mismatch for {source_label}:\n"
            f"  expected {expected}\n  actual   {actual}\n"
            "Refusing to use a binary that does not match binaries.lock.json. The "
            "download may be corrupt, truncated, or tampered with. Re-run the fetch; "
            "if it persists, the lock may be stale relative to the published artifact."
        )
    return actual


def default_cache_dir() -> Path:
    return Path.home() / ".gaia" / "agents" / "email"


@dataclass(frozen=True)
class FetchResult:
    binary_path: Path
    platform_key: str
    sha256: str
    url: str
    cached: bool


def _join_url(base: str, name: str) -> str:
    return f"{base.rstrip('/')}/{name.lstrip('/')}"


def _hub_installed_binary(platform_key: str) -> Optional[FetchResult]:
    """Return a hub-installed, checksum-verified email binary if one is present.

    Agent Hub install (#2086) writes the email agent as a platform binary into
    ``agent_install_dir("email")`` and records the hub manifest's server-computed
    SHA-256 in the ``.installed`` sentinel — the same SHA it verified the download
    against before writing. That install is authoritative: spawn it instead of
    re-gating on ``binaries.lock.json``, whose in-repo entry ships a ``PENDING-…``
    placeholder SHA until the agent is published.

    The no-unverified-binary invariant still holds — the on-disk file is re-hashed
    against the sentinel SHA and a mismatch raises loudly. Returns ``None`` (fall
    through to the lock path) when there is no binary-kind install to trust.
    """
    from gaia.hub import installer

    sentinel = installer.read_sentinel("email")
    if sentinel is None or sentinel.artifact_kind != installer.ARTIFACT_KIND_BINARY:
        return None
    if not sentinel.executable or not sentinel.artifact_sha256:
        return None
    binary_path = installer.agent_install_dir("email") / sentinel.executable
    actual = file_sha256(binary_path)
    if actual is None:
        return None
    if actual.lower() != sentinel.artifact_sha256.lower():
        raise IntegrityError(
            f"SHA-256 mismatch for the hub-installed email binary at {binary_path}:\n"
            f"  expected {sentinel.artifact_sha256}\n  actual   {actual}\n"
            "Refusing to spawn a binary that no longer matches its .installed "
            "sentinel. Reinstall the email agent from the Agent Hub."
        )
    logger.info("email sidecar: using hub-installed binary %s", binary_path)
    return FetchResult(
        binary_path, platform_key, actual, f"hub-install:{binary_path}", cached=True
    )


def fetch_binary(
    *,
    out_dir: Optional[Path] = None,
    base_url: Optional[str] = None,
    platform_key: Optional[str] = None,
    lock_path: Optional[Path] = None,
    force: bool = False,
    timeout: float = 120.0,
    session=None,
) -> FetchResult:
    """Fetch + verify + cache the email-agent binary for the current platform.

    Raises:
        PlatformError: unsupported platform / incomplete or placeholder lock entry.
        IntegrityError: SHA-256 mismatch (tampered/truncated download).
        RuntimeError: download/network failure (HTTP status surfaced).
    """
    key = platform_key or plat.current_platform_key()
    # A hub-installed, checksum-verified binary is authoritative — spawn it before
    # touching the lock, whose in-repo entry ships a placeholder SHA until publish.
    if not force:
        installed = _hub_installed_binary(key)
        if installed is not None:
            return installed

    lock = plat.load_lock(lock_path)
    entry = plat.resolve_entry(lock, key)
    resolved_base = base_url or os.environ.get("ASSETS_BASE_URL") or lock.base_url
    if not resolved_base:
        raise PlatformError(
            "no download base URL: binaries.lock.json has no baseUrl, ASSETS_BASE_URL "
            "is unset, and none was passed. Set ASSETS_BASE_URL or pass base_url."
        )
    if plat.is_placeholder_sha(entry.sha256):
        raise PlatformError(
            f"binaries.lock.json has a placeholder sha256 for '{key}' "
            f"({entry.sha256}); no binary is published for it in this build. Fetch is "
            "blocked so a bad binary can never be trusted. Publish the email agent "
            "(release_agent_email.yml) to populate the real SHA, or run dev mode "
            "(GAIA_EMAIL_AGENT_MODE=dev)."
        )

    cache = Path(out_dir) if out_dir is not None else default_cache_dir()
    cache.mkdir(parents=True, exist_ok=True)
    binary_path = cache / entry.executable
    url = _join_url(resolved_base, entry.filename)

    if not force:
        existing = file_sha256(binary_path)
        if existing and existing.lower() == entry.sha256.lower():
            logger.info("email sidecar: cache hit %s matches lock sha256", binary_path)
            return FetchResult(binary_path, key, existing, url, cached=True)

    if session is None:
        import requests

        session = requests.Session()
    logger.info("email sidecar: downloading %s binary from %s", key, url)
    resp = session.get(
        url, timeout=timeout, headers={"accept": "application/octet-stream"}
    )
    if not getattr(resp, "ok", resp.status_code == 200):
        raise RuntimeError(
            f"download failed: HTTP {resp.status_code} {getattr(resp, 'reason', '')} "
            f"for {url}. Check ASSETS_BASE_URL and that the artifact is published "
            f"for {key}."
        )
    data = resp.content
    sha = verify_sha256(data, entry.sha256, f"{key} ({url})")

    # Write to a temp then rename so a crash mid-write never leaves a
    # half-written "verified" binary. Clean up the temp on any failure.
    tmp = binary_path.with_suffix(binary_path.suffix + f".download.{os.getpid()}")
    try:
        tmp.write_bytes(data)
        os.replace(tmp, binary_path)
    except OSError:
        tmp.unlink(missing_ok=True)
        raise
    if os.name != "nt":
        binary_path.chmod(
            binary_path.stat().st_mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH
        )
    logger.info("email sidecar: installed verified binary -> %s", binary_path)
    return FetchResult(binary_path, key, sha, url, cached=False)

File: hub/agents/gaia/npm/src/fetch.ts
// Copyright(C) 2025-2026 Advanced Micro Devices, Inc. All rights reserved.
// SPDX-License-Identifier: MIT
/**
 * Binary fetcher for both components (`sidecar` and `tui`).
 *
 * Resolves the current platform → looks the component up in
 * `binaries.lock.json` → downloads it from THAT COMPONENT's base URL
 * (overridable) → **verifies its SHA-256 against the lock and fails loudly on
 * any mismatch** → writes it into a cache dir → `chmod +x` on POSIX.
 *
 * The two components come from different hub lanes: the sidecar from
 * `agents/gaia/<agentVersion>/`, the TUI from `agents/terminal-hub/<tuiVersion>/`
 * (the same binary a core install runs as `gaia tui`). Each entry's base URL
 * comes from its own component, never from a shared top-level one.
 *
 * The SHA verify is the security boundary: a tampered, truncated, or
 * not-yet-published artifact is rejected before it can ever be spawned. There is
 * NO "use it anyway" path, and a placeholder hash in the lock blocks the fetch
 * outright rather than degrading to an unverified download.
 */

import crypto from "node:crypto";
import fs from "node:fs";
import fsp from "node:fs/promises";
import os from "node:os";
import path from "node:path";

import { IntegrityError, PlatformError } from "./errors.js";
import { createLogger } from "./logger.js";
import { joinUrl } from "./url.js";
import {
  type BinaryLock,
  type BinaryLockEntry,
  type ComponentName,
  componentBaseUrl,
  currentPlatformKey,
  defaultLockPath,
  isPlaceholderSha,
  loadLock,
  resolveEntry,
} from "./platform.js";

const log = createLogger("fetch");

/**
 * Where verified binaries are cached, keyed by agent version so a version bump
 * never reuses the previous release's executables.
 */
export function defaultCacheDir(agentVersion: string): string {
  return path.join(os.homedir(), ".gaia", "npm-cache", `gaia-${agentVersion}`);
}

/**
 * The daemon's own sidecar cache — `~/.gaia/agents/gaia/`, mirroring
 * `gaia.daemon.sidecars.fetch.default_cache_dir("gaia")`. Staging the verified
 * sidecar here means the daemon's own fetch is a SHA-256 cache hit instead of a
 * second download. This path is a cross-repo contract with the daemon.
 */
export function daemonSidecarCacheDir(agentId = "gaia"): string {
  return path.join(os.homedir(), ".gaia", "agents", agentId);
}

export interface FetchOptions {
  /** Which binary to fetch. */
  component: ComponentName;
  /** Directory the verified binary is written into. Required. */
  outDir: string;
  /**
   * Override this component's `baseUrl` (e.g. a local mirror). Trailing slash
   * optional. Applies to whichever component is being fetched — the two lanes'
   * filenames never collide, so one flat mirror directory can serve both.
   */
  baseUrl?: string;
  /** Override the platform key (defaults to the current host). */
  platformKey?: string;
  /** Path to the lock file (defaults to the packaged binaries.lock.json). */
  lockPath?: string;
  /** A pre-loaded lock, to avoid re-reading it per component. */
  lock?: BinaryLock;
  /** Fetch override (tests). Defaults to global `fetch`. */
  fetchImpl?: typeof fetch;
  /** Re-download even when a verified binary is already cached. Default false. */
  force?: boolean;
  /** Abort the download after this many ms. Default 300000 (the sidecar is ~200MB). */
  timeoutMs?: number;
}

export interface FetchResult {
  component: ComponentName;
  /** Absolute path to the written, verified executable. */
  binaryPath: string;
  /** The platform key resolved. */
  platformKey: string;
  /** The verified SHA-256 (lowercase hex). */
  sha256: string;
  /** Source URL the artifact was downloaded from. */
  url: string;
  /** True when the on-disk binary was reused (hash already matched). */
  cached: boolean;
}

function sha256Hex(buf: Buffer): string {
  return crypto.createHash("sha256").update(buf).digest("hex");
}

/**
 * SHA-256 of a file on disk, or null when it is simply absent. Any other error
 * (a permission problem, a directory in the way) is re-raised: reading it as
 * "no cache" would trigger a re-download that then fails on write with a far
 * less useful message.
 */
export async function fileSha256(filePath: string): Promise<string | null> {
  try {
    return sha256Hex(await fsp.readFile(filePath));
  } catch (e) {
    const code = (e as NodeJS.ErrnoException).code;
    if (code === "ENOENT") return null;
    throw new Error(
      `cannot read the cached binary at ${filePath} to check its hash: ${(e as Error).message}. ` +
        "Fix the permissions on that path, or pass a different --cache-dir / --sidecar-dir.",
      { cause: e },
    );
  }
}

/**
 * Verify a buffer against an expected SHA-256. Throws `IntegrityError` loudly on
 * mismatch — this is the no-silent-fallback integrity gate.
 */
export function verifySha256(
  buf: Buffer,
  expected: string,
  sourceLabel: string,
): string {
  const actual = sha256Hex(buf);
  if (actual.toLowerCase() !== expected.toLowerCase()) {
    throw new IntegrityError(
      `SHA-256 mismatch for ${sourceLabel}:\n` +
        `  expected ${expected}\n` +
        `  actual   ${actual}\n` +
        "Refusing to use a binary that does not match binaries.lock.json. " +
        "The download may be corrupt, truncated, or tampered with. Re-run the fetch; " +
        "if it persists, the lock is stale relative to the published artifact — " +
        "report it at https://github.com/amd/gaia/issues.",
    );
  }
  return actual;
}

/**
 * Fetch + verify + install one component's binary for the current platform.
 *
 * @throws PlatformError   unsupported platform / incomplete entry / placeholder hash
 * @throws IntegrityError  SHA-256 mismatch
 * @throws Error           download/network failure (HTTP status surfaced)
 */
export async function fetchBinary(opts: FetchOptions): Promise<FetchResult> {
  if (!opts?.outDir) {
    throw new TypeError("fetchBinary requires an outDir to write the binary into");
  }
  if (!opts?.component) {
    throw new TypeError('fetchBinary requires a component ("sidecar" or "tui")');
  }
  const fetchImpl = opts.fetchImpl ?? globalThis.fetch;
  if (typeof fetchImpl !== "function") {
    throw new TypeError("global fetch unavailable — use Node >= 18 or pass fetchImpl");
  }

  const lock: BinaryLock = opts.lock ?? loadLock(opts.lockPath ?? defaultLockPath());
  const platformKey = opts.platformKey ?? currentPlatformKey();
  const entry: BinaryLockEntry = resolveEntry(lock, opts.component, platformKey);
  // Per-component: the sidecar and the TUI live in different hub lanes at
  // different versions, so there is no single base URL to fall back to.
  const baseUrl = opts.baseUrl ?? componentBaseUrl(lock, opts.component);

  if (isPlaceholderSha(entry.sha256)) {
    throw new PlatformError(
      `binaries.lock.json has a placeholder sha256 for ${opts.component}/'${platformKey}' ` +
        `(${entry.sha256}), so no verifiable binary is published for it in this build. ` +
        "Fetch is blocked so an unverifiable binary can never be trusted. " +
        "Install a released @amd-gaia/gaia, or build the binary locally " +
        "(hub/agents/gaia/python/packaging for the sidecar, `make -C tui cross-compile` " +
        "for the TUI) and point the lifecycle helpers at it directly.",
    );
  }

  const outDir = path.resolve(opts.outDir);
  await fsp.mkdir(outDir, { recursive: true });
  const binaryPath = path.join(outDir, entry.executable);
  const url = joinUrl(baseUrl, entry.filename);

  if (!opts.force) {
    const existing = await fileSha256(binaryPath);
    if (existing && existing.toLowerCase() === entry.sha256.toLowerCase()) {
      log.debug(`cache hit: ${binaryPath} already matches lock sha256`);
      // Re-apply the exec bit: an interrupted earlier run or a restrictive umask
      // can leave correct bytes that still cannot be spawned.
      if (process.platform !== "win32") await fsp.chmod(binaryPath, 0o755);
      return {
        component: opts.component,
        binaryPath,
        platformKey,
        sha256: existing,
        url,
        cached: true,
      };
    }
  }

  log.info(`downloading ${opts.component} for ${platformKey} from ${url}`);
  const timeoutMs = opts.timeoutMs ?? 300_000;
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  let buf: Buffer;
  try {
    const res = await fetchImpl(url, {
      headers: { accept: "application/octet-stream" },
      signal: controller.signal,
    });
    if (!res.ok) {
      throw new Error(
        `download failed: HTTP ${res.status} ${res.statusText} for ${url}. ` +
          "Check the base URL and that the artifact is published for this platform.",
      );
    }
    buf = Buffer.from(await res.arrayBuffer());
  } catch (e) {
    if ((e as Error).name === "AbortError") {
      throw new Error(`download timed out after ${timeoutMs}ms for ${url}`);
    }
    throw e;
  } finally {
    clearTimeout(timer);
  }
  log.debug(`downloaded ${buf.length} bytes`);

  const sha = verifySha256(buf, entry.sha256, `${opts.component} ${platformKey} (${url})`);

  // Write to a temp then rename, so a crash mid-write never leaves a
  // half-written file that a later run would treat as a real binary.
  const tmp = `${binaryPath}.download.${process.pid}`;
  try {
    await fsp.writeFile(tmp, buf);
    await fsp.rename(tmp, binaryPath);
  } catch (e) {
    await fsp.rm(tmp, { force: true }).catch(() => undefined);
    throw e;
  }

  if (process.platform !== "win32") {
    await fsp.chmod(binaryPath, 0o755);
  }

  log.info(`installed verified ${opts.component} -> ${binaryPath}`);
  return {
    component: opts.component,
    binaryPath,
    platformKey,
    sha256: sha,
    url,
    cached: false,
  };
}

export interface FetchAllOptions extends Omit<FetchOptions, "component" | "outDir"> {
  /** Cache dir for the TUI. Defaults to `defaultCacheDir(lock.agentVersion)`. */
  tuiDir?: string;
  /** Cache dir for the sidecar. Defaults to the daemon's `~/.gaia/agents/gaia`. */
  sidecarDir?: string;
}

export interface FetchAllResult {
  sidecar: FetchResult;
  tui: FetchResult;
  lock: BinaryLock;
}

/**
 * Fetch + verify BOTH binaries. Sequential on purpose: the sidecar is the large
 * download and a failure there should not race a half-finished TUI download.
 *
 * The sidecar lands in the daemon's own cache dir so the daemon — which owns
 * spawning it — finds it already verified instead of downloading it again.
 */
export async function fetchAll(opts: FetchAllOptions = {}): Promise<FetchAllResult> {
  const lock = opts.lock ?? loadLock(opts.lockPath ?? defaultLockPath());
  const platformKey = opts.platformKey ?? currentPlatformKey();
  const common = { ...opts, lock, platformKey };
  const sidecar = await fetchBinary({
    ...common,
    component: "sidecar",
    outDir: opts.sidecarDir ?? daemonSidecarCacheDir(),
  });
  const tui = await fetchBinary({
    ...common,
    component: "tui",
    outDir: opts.tuiDir ?? defaultCacheDir(lock.agentVersion),
  });
  return { sidecar, tui, lock };
}

/** Sync existence check, for the lifecycle layer. */
export function binaryExists(binaryPath: string): boolean {
  return fs.existsSync(binaryPath);
}
